Develop q833

ErikWegner opened the pull request.

ErikWegner committed:
Typo

ErikWegner committed:
A test to branch to a conditional transition.

ErikWegner committed:
The RepositoryBasedTransitionHandler can access instances to start another workflow.

dmolineus commented on README.md:
```diff
@@ -66,7 +66,7 @@ Example
 
 A concrete implementation is available as integration for CMS Contao 
-[netzmacht/contao-workflow](https:github.com/netzmacht/contao-workflow).
+[netzmacht/contao-workflow](https://github.com/netzmacht/contao-workflow).
```
Nice catch :+1:

dmolineus reviewed (changes requested):
Hallo Erik,

danke soweit für deine Vorarbeit. Bei der Workflow-Bibliothek geht es mir darum die Features zu beschränken, sodass es als Basis-Bibliothek für Workflow-Anwendungen genutzt werden kann.

Daher möchte ich hier das Feature der bedingten Transitions nicht drin haben, da es für mich keine generische Anforderung ist. Mein Ansatz wäre gewesen hierfür eine eigene Action zu bauen und die als PostAction einer Transition hinzuzufügen.

Eine mögliche Implementierung wäre:

```php

use Netzmacht\Workflow\Flow\Action;
use Netzmacht\Workflow\Flow\Context;
use Netzmacht\Workflow\Flow\Context\ErrorCollection;
use Netzmacht\Workflow\Flow\Context\Properties;
use Netzmacht\Workflow\Flow\Exception\ActionFailedException;
use Netzmacht\Workflow\Flow\Item;
use Netzmacht\Workflow\Flow\Transition;

final class ConditionalTransitionsAction implements Action
{
    /** @var Transition[] */
    private $transitions = [];

    public function getRequiredPayloadProperties(Item $item): array
    {
        // We check the pre conditions which may not depend on the context, so let's create an empty context here
        $context = new Context(new Properties(), new Properties(), new ErrorCollection());
        $transition = $this->getAllowedTransition($item, $context);
        
        if ($transition === null) {
            return [];
        }
        
        return $transition->getRequiredPayloadProperties($item);
    }

    public function validate(Item $item, Context $context): bool
    {
        $transition = $this->getAllowedTransition($item, $context);
        if ($transition === null) {
            // No transition matched, the transition is not executable
            // TODO: Add error to the context

            return false;
        }

        if (! $transition->isAvailable($item, $context)) {
            // TODO: Add error to the context
            return false;
        }

        return $transition->validate($item, $context);
    }

    public function transit(Transition $transition, Item $item, Context $context): void
    {
        $ConditionalTransition = $this->getAllowedTransition($item, $context);
        if ($ConditionalTransition === null) {
            throw ActionFailedException::action($this, $context->getErrorCollection());
        }

        $ConditionalTransition->execute($item, $context);
    }

    private function getAllowedTransition(Item $item, Context $context): ?Transition
    {
        foreach ($this->transitions as $transition) {
            if ($transition->isAllowed($item, $context)) {
                return $transition;
            }
        }

        return null;
    }
}
```

Als Action kann sie meinetwegen auch Teil der Bilbiothek bleiben. Andernfalls (als Teil der Transition) bitte in contao-workflow to Transition-Klasse extenden und dort implementieren.

dmolineus commented:
Bzgl. des Übergangs zu einem neuen Workflow habe ich folgende Ansätze überlegt. Alle haben für mich Vor- und Nachteile:

**Verknüpfung der Start-Transition des neuen Workflows**

 - Bedingungen der Start-Transition könnten nicht erfüllt sein, wenn der Workflow bereits existiert

**Gleichzeitiges Ausführen beider Transitions im Transition Handler**

 - Lässt sich nicht sauber abbilden, da nicht sicher gestellt werden kann, dass vorab schon bekannt ist ob der neue Workflow gestartet werden kann

**Workflow über Action starten**

 - Ähnlich wie bei den Conditional Transitions
 - Bedingungen der Start-Transition könnten nicht erfüllt sein, wenn der Workflow bereits existiert

**Zielschritt der Transition verweist auf einen Schritt eines anderen Workflows**

 - Derzeit hat der Schritt keine Information zum Workflow `workflowName`, dies müsste dann sichergestellt werden
 - Es gibt nicht mehr einen initialien Startprozess des Workflows

Ich tendiere gerade zu letzterem. Vielleicht ergibt es Sinn hier uns nochmal telefonisch zu verständigen.

Als Denkansatz mein Versuch der Implementierung des angedachten Dekorators:

```php
use Netzmacht\Workflow\Flow\Context;
use Netzmacht\Workflow\Flow\Exception\FlowException;
use Netzmacht\Workflow\Flow\Item;
use Netzmacht\Workflow\Flow\State;
use Netzmacht\Workflow\Flow\Step;
use Netzmacht\Workflow\Flow\Transition;
use Netzmacht\Workflow\Flow\Workflow;
use Netzmacht\Workflow\Handler\TransitionHandler;
use Netzmacht\Workflow\Manager\WorkflowManager;

class ChangeWorkflowTransitionHandler implements TransitionHandler
{
    /** @var TransitionHandler */
    private $transitionHandler;

    /** @var TransitionHandler|null */
    private $nextWorkflowTransitionHandler;

    /** @var WorkflowManager */
    private $workflowManager;

    /** @var Step|null */
    private $expectedStepTo;

    public function __construct(WorkflowManager $workflowManager, TransitionHandler $transitionHandler)
    {
        $this->transitionHandler = $transitionHandler;
        $this->workflowManager = $workflowManager;
        $this->expectedStepTo = $transitionHandler->getTransition()->getStepTo();

        if ($this->expectedStepTo && $this->expectedStepTo->startsWorkflow()) {
            $workflow = $workflowManager->getWorkflowByName($this->expectedStepTo->getNextWorkflowName());

            // Method createTransitionHandler does not yet exist, should extract the factory out
            // of workflowManager::handle
            $this->nextWorkflowTransitionHandler = $this->workflowManager->createTransitionHandler(
                $workflow,
                $this->getItem()
            );
        }
    }

    public function getWorkflow(): Workflow
    {
        return $this->transitionHandler->getWorkflow();
    }

    public function getItem(): Item
    {
        return $this->transitionHandler->getItem();
    }

    public function getTransition(): Transition
    {
        return $this->transitionHandler->getTransition();
    }

    public function getCurrentStep(): ?Step
    {
        return $this->transitionHandler->getCurrentStep();
    }

    public function isWorkflowStarted(): bool
    {
        return $this->transitionHandler->isWorkflowStarted();
    }

    public function getRequiredPayloadProperties(): array
    {
        if (!$this->nextWorkflowTransitionHandler) {
            return $this->transitionHandler->getRequiredPayloadProperties();
        }

        return array_merge(
            $this->transitionHandler->getRequiredPayloadProperties(),
            $this->nextWorkflowTransitionHandler->getRequiredPayloadProperties()
        );
    }

    public function isAvailable(): bool
    {
        if (! $this->transitionHandler->isAvailable()) {
            return false;
        }

        if ($this->nextWorkflowTransitionHandler) {
            return $this->nextWorkflowTransitionHandler->isAvailable();
        }

        return true;
    }

    public function getContext(): Context
    {
        return $this->transitionHandler->getContext();
    }

    public function validate(array $payload = []): bool
    {
        if (!$this->transitionHandler->validate($payload)) {
            return false;
        }

        if ($this->nextWorkflowTransitionHandler) {
            return $this->nextWorkflowTransitionHandler->validate($payload);
        }

        return true;
    }

    public function transit(): State
    {
        $state = $this->transitionHandler->transit();

        if (!$state->isSuccessful() || !$this->nextWorkflowTransitionHandler) {
            return $state;
        }

        if ($this->expectedStepTo->getName() !== $state->getStepName()) {
            throw new FlowException('Unexpected step reached');
        }

        return $this->nextWorkflowTransitionHandler->transit();
    }
}
```

ErikWegner committed:
A failing test for the new behaviour to come

ErikWegner committed:
Test: A post action can execute a transition.

ErikWegner committed:
Moving implementation to contao-workflow

ErikWegner committed:
ChangeWorkflowTransitionHandler as decorator

ErikWegner committed:
First implementation of ChangeWorkflowTransitionHandler

ErikWegner committed:
Fix tests

ErikWegner committed:
WorkflowManager.getWorkflowByItem takes running workflow into account
ChangeWorkflowTransitionHandler modified to finally support changing the workflow

dmolineus commented on src/Handler/ChangeWorkflowTransitionHandler.php:
```diff
@@ -0,0 +1,143 @@
+    {
+        // TODO: handle $nextWorkflowTransitionHandler->getRequiredPayloadProperties
+        if ($this->executeWorkflowTransition) {
+        }
```
The todo task is not resolved

dmolineus commented on src/Handler/ChangeWorkflowTransitionHandler.php:
```diff
@@ -0,0 +1,143 @@
+    {
+        // TODO: handle $nextWorkflowTransitionHandler->getRequiredPayloadProperties
+        if ($this->executeWorkflowTransition) {
+
```
The todo task is not resolved.

dmolineus commented on src/Manager/Manager.php:
```diff
@@ -46,6 +46,19 @@ interface Manager
+     * @return TransitionHandler
+     */
+    public function createTransitionHandler(Workflow $workflow, Item $item, string $transitionName, bool $changeWorkflow): TransitionHandler;
+
```
Adding a new method is a BC break. The goal was and is to avoid a new major version.

In addition the `CachedManager` doesn't implement the method

dmolineus commented on src/Handler/ChangeWorkflowTransitionHandler.php:
```diff
@@ -0,0 +1,143 @@
+
+namespace Netzmacht\Workflow\Handler;
+
+use Netzmacht\ContaoWorkflowBundle\Workflow\Definition\Loader\WorkflowLoader;
```
Class does not contain to this library

dmolineus commented on src/Handler/ChangeWorkflowTransitionHandler.php:
```diff
@@ -0,0 +1,143 @@
+ * @copyright  2014-2017 netzmacht David Molineus
+ * @license    LGPL 3.0
+ * @filesource
+ */
```
Wrong header snippet used

dmolineus reviewed (commented):
The implemented approach using a `ChangeWorkflowTransitionHandler` wasn't implemented in a workable state. I decided to use the proposed action based solution instead
